public/app/features/plugins/extensions/getPluginExtensions.ts

```diff
@@ -229,3 +278,60 @@ function getLinkExtensionPathWithTracking(pluginId: string, path: string, config
     })
   );
 }
+
+export function getSidecarHelpers(
+  dispatch: AppDispatch,
+  getSidecarAppId: () => string | undefined,
+  locationService: LocationService
+): SidecarHelpers {
+  const closeApp = (appId: string) => {
+    if (!config.featureToggles.appSidecar) {
+      console.warn('App sidecar feature toggle is not enabled, doing nothing');
+      return;
+    }
+    dispatch(appSidecarSlice.actions.closeApp({ appId }));
+  };
+
+  const openApp = (appId: string) => {
+    if (!config.featureToggles.appSidecar) {
+      console.warn('App sidecar feature toggle is not enabled, doing nothing');
+      return;
+    }
+    dispatch(appSidecarSlice.actions.openApp({ appId }));
+  };
+
+  const getOpenedApps = () => {
+    if (!config.featureToggles.appSidecar) {
+      console.warn('App sidecar feature toggle is not enabled, doing nothing');
+      return [];
+    }
+
+    const openedApps = [];
+    openedApps.push(getMainAppId(locationService.getLocation().pathname));
+    const sidecarAppId = getSidecarAppId();
+    if (sidecarAppId) {
+      openedApps.push(sidecarAppId);
+    }
+    return openedApps;
+  };
+
+  return {
+    openApp,
+    closeApp,
+    getOpenedApps,
+  };
+}
+
+function getMainAppId(pathname: string) {
+  // A naive way to sort of simulate core features being an app and having an appID
+  let mainApp = pathname.match(/\/a\/([^/]+)/)?.[1];
+  if (!mainApp && pathname.match(/\/explore/)) {
+    mainApp = 'explore';
+  }
+
+  if (!mainApp && pathname.match(/\/d\//)) {
+    mainApp = 'dashboards';
+  }
+
+  return mainApp || 'unknown';
+}
```

Comment:
Hmm, I think this can get even trickier - if an app plugin is registering itself (or any of its pages) under somewhere else in the nav tree, then the URL won't tell you if the page is served by an app plugin.

Where/why would this be important to know? To not open the app in a side-view if it is already visible for the user?

Reply:
> if an app plugin is registering itself (or any of its pages) under somewhere else in the nav tree, then the URL won't tell you if the page is served by an app plugin.

Ah didn't know that was possible.

So the reason for this is for example if Explore created a link extension for Query Library, that extension could change depending of whether query library is shown next to Explore or not. Like right now it could change from "Run query" button because you are already showing Explore, or to `Open in explore` in case we show Query Library next to dashboard.

I assume this would be easier where explore, dashboards etc would be implemented as app plugins so this just tries to give that information somehow anyway.

Imho at this point it does not have to be perfect. Like this code isn't all the demo code anyway and that is why it's behind feature flag and experimental and I don't want to create too much changes to be reviewed at once.

Reply:
Yeah, I think this solution should be good enough for now. As long as we pass `isAppOpened` to the extensions as a boolean, we can always fix / tune / refactor the underlying logic seamlessly under the hood.